python/CLOUDTRAIL_S3_DATAEVENTS_ENABLED/CLOUDTRAIL_S3_DATAEVENTS_ENABLED.py

```diff
@@ -0,0 +1,488 @@
+     Then: Return NON_COMPLIANT
+  Scenario: 7
+    Given: Rule parameter is provided and valid
+      And: At least one AWS CloudTrail trail exists
+      And: At least one of the trails 'HasCustomEventSelectors' set as 'True'
+      And: The Type in 'DataResources' returned from 'get_event_selectors' is AWS::S3::Object
+      And: The Value in 'DataResources' returned from 'get_event_selectors' includes the S3 bucket as specified in rule parameter 'S3BucketName'.
+     Then: Return COMPLIANT
+  Scenario: 8
+    Given: Rule parameter value is empty
+      And: At least one AWS CloudTrail trail exists
+      And: At least one of the trails 'HasCustomEventSelectors' set as 'True'
+      And: The Type in 'DataResources' returned from 'get_event_selectors' is AWS::S3::Object
+      And: The Value in 'DataResources' returned from 'get_event_selectors' is arn:aws:s3:::
+     Then: Return COMPLIANT
+'''
+import json
+import sys
+import datetime
+import re
+import boto3
+import botocore
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+DEFAULT_CHECK = 'arn:aws:s3'
+
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    ct_client = get_client('cloudtrail', event)
+    evaluations = []
+    compliant_buckets = []
+    non_compliant_buckets = []
+    trail_list = get_all_trails(ct_client)
+    region = event["configRuleArn"][15:].split(':')[0]
+
+    #No trails configured: return NON_COMPLIANT
+    if not trail_list:
+        evaluations.append(build_evaluation(event['accountId'], 'NON_COMPLIANT', event, annotation='No AWS CloudTrail Trails are present to log events in this region.'))
+        return evaluations
+
+    #No rule parameter specified, default check for all S3 buckets
+    if not valid_rule_parameters:
+        for trail in trail_list:
+            trail_region = trail['HomeRegion']
+            if trail_region == region:
+                if trail['HasCustomEventSelectors']:
+                    event_selector_output = ct_client.get_event_selectors(TrailName=trail['Name'])['EventSelectors']
+                    data_resource_values = get_data_resource_s3_values(event_selector_output)
+                    if isinstance(data_resource_values, bool):
+                        evaluations.append(build_evaluation(event['accountId'], 'COMPLIANT', event))
+                        return evaluations
+
+    #Rule parameter S3BucketName is specified
+    else:
+        for trail in trail_list:
+            trail_region = trail['HomeRegion']

```

Comment:
We might have been on the wrong path here (apologies for that). It might be confusing for the customer to consider s3 buckets as parameters in different regions. Especially if they have a multi-region trail on all s3 buckets.

I see that we can filter shadow trails on-demand. Having an input parameter IncludeShadowTrail (optional) with a default to true, might be easier. 

You can then describe the appropriate trails on-demand with a filter on the DescribeTrails API call and not anymore care about the HomeRegion.

What do you think?

 https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_DescribeTrails.html

Reply:
Not sure if this will affect the code logic as IncludeShadowTrail is set to True by default, we get all trails and need to check for HomeRegion. 
If set to False, we won't get other region trails but if we add this as a parameter, we will have to check for it's value and then decide whether to check for HomeRegion or not.

Basically, we would be removing one check and adding another.

We can start as default of False, this will limit the number of trails returned and reduce the output of API call.